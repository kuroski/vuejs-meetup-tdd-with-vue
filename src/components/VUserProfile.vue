<script>
export default {
  name: 'user-profile',
  props: {
    user: {
      type: Object,
      required: true,
    }
  }
}
</script>

<template>
  <div>
    <img :src="user.avatar_url" />
    <div class="name">
      {{ user.name }}
    </div>
  </div>
</template>

