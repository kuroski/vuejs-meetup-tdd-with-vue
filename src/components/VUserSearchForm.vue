<script>
export default {
  name: 'user-search-form',
  props: {
    search: {
      type: Function,
      required: true,
    },
  },
  data() {
    return {
      term: '',
    }
  },
}
</script>

<template>
  <form @submit.prevent="search(term)">
    <input type="text" v-model="term" />
    <button type="submit">Pesquisar</button>
  </form>
</template>
