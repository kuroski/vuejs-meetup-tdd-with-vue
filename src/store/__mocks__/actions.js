export default {
  SEARCH_USER: jest.fn(),
}