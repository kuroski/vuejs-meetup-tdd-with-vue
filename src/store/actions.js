import api from '@/api'

export default {
  SEARCH_USER({ commit }, { username }) {
    return new Promise(async (resolve, reject) => {
      try {
        const user = await api.searchUser(username)
        commit('SET_USER', user)
        resolve(user)
      } catch(error) {
        reject(error)
      }
    })
  }
}