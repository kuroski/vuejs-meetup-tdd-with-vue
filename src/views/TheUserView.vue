<script>
import { mapState } from 'vuex'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'

export default {
  name: 'user-view',
  components: {
    VUserSearchForm,
    VUserProfile,
  },
  computed: {
    ...mapState({
      user: 'user'
    })
  },
  methods: {
    searchUser(username) {
      this.$store.dispatch('SEARCH_USER', { username })
    }
  }
}
</script>

<template>
  <div>
    <VUserSearchForm :search="searchUser" />
    <VUserProfile :user="user" />
  </div>
</template>

