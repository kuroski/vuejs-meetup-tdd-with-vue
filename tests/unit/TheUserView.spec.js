jest.mock('@/store/actions')
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import state from '@/store/state'
import actions from '@/store/actions'
import TheUserView from '@/views/TheUserView'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'
import user from './fixtures/user'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('TheUserView', () => {

  const build = () => {
    const store = new Vuex.Store({
      state,
      actions,
    })

    const wrapper = shallowMount(TheUserView, {
      localVue,
      store,
    })

    return {
      wrapper,
      userSearchForm: () => wrapper.find(VUserSearchForm),
      userProfile: () => wrapper.find(VUserProfile),
    }
  }

  it('renders the view', () => {
    // arranje
    const { wrapper } = build()
    // assert
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders main view components', () => {
    // arranje
    const { userSearchForm, userProfile } = build()
    // assert
    expect(userSearchForm().exists()).toBe(true)
    expect(userProfile().exists()).toBe(true)
  })

  it('passes a binded search function to user search form component', () => {
    // arranje
    const { wrapper, userSearchForm } = build()
    // assert
    expect(userSearchForm().vm.search).toBe(wrapper.vm.searchUser)
  })

  it('passes a binded user prop to user profile component', () => {
    // arranje
    state.user = user.success
    const { wrapper, userProfile } = build()
    // assert
    expect(wrapper.vm.user).toBe(state.user)
    expect(userProfile().vm.user).toBe(wrapper.vm.user)
  })

  it('searches for a user', () => {
    // arranje
    const expectedUser = 'kuroski'
    const { wrapper } = build()
    // act
    wrapper.vm.searchUser(expectedUser)
    // assert
    expect(actions.SEARCH_USER).toHaveBeenCalled()
    expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual({ username: expectedUser })
  })
})