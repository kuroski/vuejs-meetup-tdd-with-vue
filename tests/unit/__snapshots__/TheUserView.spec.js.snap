// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheUserView renders the view 1`] = `
<div>
  <!---->
  <!---->
</div>
`;
